### PR TITLE
Feat/select v 2

### PR DESCRIPTION
**Pull request by hirsch88**

### CODE REVIEW

File: packages/library/src/components/bal-select-option/bal-select-option.scss [hunk 1]
```diff
@@ -2,9 +2,12 @@
 @import 'node_modules/bulma/sass/components/dropdown.sass';
 
 :host {
-  display: block;
   font-family: $family-primary !important;
 
+  &.is-connected {
+    display: block;
```
Comment (dweber019) [inline packages/library/src/components/bal-select-option/bal-select-option.scss:8]: May using `initial`

File: packages/library/src/components/bal-select-option/bal-select-option.tsx [hunk 2]
```diff
@@ -14,25 +18,27 @@ export class SelectOption implements ComponentInterface {
...
   /**
-   * If `true` the option is hidden
+   * If `true` the option has a checkbox
    */
-  @Prop({ reflect: true }) hidden = true
+  @Prop() checkbox = false
 
   /**
-   * Baloise icon as a prefix
+   * The value of the dropdown item. This value will be returned by the parent `<bal-dropdown>` element.
```
Comment (dweber019) [inline packages/library/src/components/bal-select-option/bal-select-option.tsx]: Strange as we have here a `bal-select-option` => `bal-select` and here you talk about `<bal-dropdown>`

File: packages/library/src/components/bal-select/bal-select.scss [hunk 3]
```diff
@@ -5,17 +5,122 @@
...
+      display: flex;
+      flex: 1 1;
+      flex-wrap: wrap;
+      line-height: 18px;
+      max-width: 100%;
+      min-width: 0;
+    }
+
+    bal-tag {
+      margin-right: 8px;
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.scss:37]: do we have a standard padding/marging?

File: packages/library/src/components/bal-select/bal-select.scss [hunk 4]
```diff
@@ -5,17 +5,122 @@
...
+    }
+  }
 
-.multiple-typeahead {
-  padding: 15px 20px;
+  .bal-select__empty {
+    padding: 10px 20px 13px 20px;
+  }
+
+  // .bal-select-input {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.scss]: If you don't need it, remove it.

File: packages/library/src/components/bal-select/option.utils.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import { BalOptionController } from '../bal-select-option/bal-select-option'
+
+export function addOption(options: BalOptionController[], option: BalOptionController): BalOptionController[] {
+  return [...options, option]
+}
+
+export function updateOption(options: BalOptionController[], optionToUpdate: BalOptionController): BalOptionController[] {
+  const copyOfOptions = [...options]
+  for (let index = 0; index < copyOfOptions.length; index++) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/option.utils.ts]: Why not map?
=> returns new array
if (option.value === optionToUpdate.value) {
  return optionToUpdate
}
return option;

File: packages/library/src/components/bal-select/option.utils.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    if (option.value === optionToUpdate.value) {
+      copyOfOptions[index] = optionToUpdate
+    }
+  }
+  return copyOfOptions
+}
+
+export function removeOption(options: BalOptionController[], optionToRemove: BalOptionController): BalOptionController[] {
+  const copyOfOptions = []
+  for (let index = 0; index < options.length; index++) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/option.utils.ts]: Same here with map

File: packages/library/src/components/bal-select/option.utils.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    if (option.value !== optionToRemove.value) {
+      copyOfOptions.push(option)
+    }
+  }
+  return copyOfOptions
+}
+
+export function findOptionByValue(options: BalOptionController[], value: string): BalOptionController | undefined {
+  if (options.length > 0) {
+    for (let index = 0; index < options.length; index++) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/option.utils.ts]: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/find

File: packages/library/src/components/bal-select/option.utils.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    }
+  }
+
+  return undefined
+}
+
+export function findOptionByLabel(options: BalOptionController[], label: string): BalOptionController | undefined {
+  if (options.length > 0) {
+    for (let index = 0; index < options.length; index++) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/option.utils.ts]: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/find

File: packages/library/src/components/bal-select/value.utils.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import { BalOptionController } from '../bal-select-option/bal-select-option'
+import { findOptionByLabel, findOptionByValue } from './option.utils'
+
+export function removeValue(value: string[], valueToRemove: string): string[] {
+  const copyOfValues = []
+  for (let index = 0; index < value.length; index++) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/value.utils.ts]: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

File: packages/library/src/components/bal-select/value.utils.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      copyOfValues.push(value[index])
+    }
+  }
+  return copyOfValues
+}
+
+export function addValue(values: string[], valueToAdd: string, hasMultipleValue: boolean) {
+  if (hasMultipleValue) {
+    if (values.includes(valueToAdd)) {
+      return removeValue(values, valueToAdd)
```
Comment (dweber019) [inline packages/library/src/components/bal-select/value.utils.ts:11]: Strange as the method is called `addValue` but you remove values here?
Is it a toggle?

File: packages/library/src/components/bal-select/value.utils.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+export function compareValueWithInput(text: string, input: string): boolean {
+  text = text.toLocaleLowerCase()
+  input = input.toLocaleLowerCase()
+  return text.indexOf(input) >= 0
+}
+
+export function validateAfterBlur(values: string[], options: BalOptionController[], typedLabel: string): string[] {
+  if (values.length > 0) {
+    const valueOption = findOptionByValue(options, values[0])
+    if (valueOption && typedLabel !== valueOption?.label) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/value.utils.ts:27]: As you have checked `valueOption` you don't need `?`, right?

File: packages/library/src/components/bal-select/bal-select.tsx [hunk 12]
```diff
@@ -218,300 +189,241 @@ export class Select {
...
-    if (this.isDropdownOpen && (isArrowDownKey(event) || isArrowUpKey(event))) {
-      event.stopPropagation()
-      event.preventDefault()
+    if (!this.multiple) {
+      if (this.typeahead) {
+        this.filterOptions(selectedOption.label)
+      }
+
+      await this.dropdownElement.close()
+    } else {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.tsx:246]: else can be removed if you return the `await`s

File: packages/library/src/components/bal-select/bal-select.tsx [hunk 13]
```diff
@@ -218,300 +189,241 @@ export class Select {
...
+   ********************************************************/
 
-    if (this.typeahead && !this.multiple) {
-      this.clearTimeout = setTimeout(() => {
-        const hasInputValueOptionLabel = this.childOptions.some(option => this.compareLabels(option.label || '', this.inputElementValue))
-        if (!hasInputValueOptionLabel) {
-          this.clear()
-        }
-      }, 500)
+  private async updateFocus() {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.tsx]: async can be removed

File: packages/library/src/components/bal-select/bal-select.tsx [hunk 14]
```diff
@@ -218,300 +189,241 @@ export class Select {
...
+    } else {
+      this.focusIndex = 0
     }
   }
 
-  private preventEvent = (ev: Event) => {
-    ev.preventDefault()
-    ev.stopPropagation()
-  }
+  private async scrollToFocusedOption(focusedElement?: HTMLBalSelectOptionElement) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.tsx]: async can be removed

File: packages/library/src/components/bal-select/bal-select.tsx [hunk 15]
```diff
@@ -218,300 +189,241 @@ export class Select {
...
     }
   }
 
-  render() {
-    const labelId = this.inputId + '-lbl'
-    const label = findItemLabel(this.el)
-    if (label) {
-      label.id = labelId
-      label.htmlFor = this.inputId
+  private async scrollTo(scrollTop: number) {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.tsx]: async can be removed

File: packages/library/src/components/bal-select/bal-select.tsx [hunk 16]
```diff
@@ -218,300 +189,241 @@ export class Select {
...
     }
+  }
 
-    this.scrollToFocusedOption()
+  private focusOptionByLabel(key: string) {
+    this.labelToScrollTo = this.labelToScrollTo + key
+    clearTimeout(this.clearScrollToValue)
+    this.clearScrollToValue = setTimeout(() => {
+      this.scrollToLabel(this.labelToScrollTo)
+    }, 600)
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.tsx:404]: can this be done quicker or without a timeout?

File: packages/library/src/components/bal-select/bal-select.tsx [hunk 17]
```diff
@@ -520,130 +432,222 @@ export class Select {
...
+        this.filterOptions(this.inputElement.value)
+      }
+    }
+    event.stopPropagation()
   }
 
-  private focusOptionElement(optionElement: HTMLBalSelectOptionElement) {
-    this.unfocusAllOptions()
-    optionElement.focused = true
+  private handleInputBlur = async (event: FocusEvent) => {
```
Comment (dweber019) [inline packages/library/src/components/bal-select/bal-select.tsx]: is async needed

**Review by dweber019 — CHANGES_REQUESTED** [review]

File: examples/angular/src/app/form/form.component.ts [hunk 18]
```diff
@@ -24,7 +24,8 @@ export class FormComponent {
   }
 
   setCommentDisabled() {
-    this.form.get('comment')?.disable()
+    console.log(this.form.get('age'))
```
Comment (mladenplaninicic) [inline examples/angular/src/app/form/form.component.ts]: should we remove this comment?

File: packages/library/src/components/bal-select/bal-select.tsx (hunk 16 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.tsx:404]: no this is the delay for the user to finish his search value

File: packages/library/src/components/bal-select-option/bal-select-option.scss (hunk 1 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select-option/bal-select-option.scss:8]: this class is added after the component is connected

File: packages/library/src/components/bal-select-option/bal-select-option.tsx (hunk 2 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select-option/bal-select-option.tsx]: done

File: packages/library/src/components/bal-select/bal-select.scss (hunk 3 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.scss:37]: the spacing guide is under development

File: packages/library/src/components/bal-select/bal-select.scss (hunk 4 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.scss]: done

File: packages/library/src/components/bal-select/option.utils.ts (hunk 5 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/option.utils.ts]: used map

File: packages/library/src/components/bal-select/bal-select.tsx (hunk 13 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.tsx]: done

File: packages/library/src/components/bal-select/bal-select.tsx (hunk 14 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.tsx]: done

File: packages/library/src/components/bal-select/bal-select.tsx (hunk 15 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.tsx]: done

File: packages/library/src/components/bal-select/bal-select.tsx (hunk 17 above)
Comment (hirsch88) [reply packages/library/src/components/bal-select/bal-select.tsx]: done

File: examples/angular/src/app/form/form.component.ts (hunk 18 above)
Comment (hirsch88) [reply examples/angular/src/app/form/form.component.ts]: removed

**Review by dweber019 — APPROVED** [review]

**Pull request merged**